### ISSUE TITLE
hibernate-validator + kubernetes-client native-image is slow and resource intensive

### ISSUE THREAD

**Issue by jorsol** [quarkusio/quarkus#3593]
**Describe the bug**
Including the quarkus-hibernate-validator on a fairly dummy project makes the compilation time to last the double or triple, and increase the CPU and Memory usage.

On an idle and relatively powerful workstation (AMD Ryzen 7 2700X, 64GB RAM), the compilation time without quarkus-hibernate-validator in three runs:
01:20 min
01:23 min
01:21 min

Creating a final file size of 39 MB.

After adding the quarkus-hibernate-validator the times go up to:
03:48 min
03:38 min
03:38 min

Creating a final file size of 73 MB. * Just by adding hibernate-validator

**To Reproduce**
1. `./mvnw clean package -Pnative` without quarkus-hibernate-validator
2. ./mvnw quarkus:add-extension -Dextensions="quarkus-hibernate-validator"
3. `./mvnw clean package -Pnative`

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: Linux ibz 5.0.0-25-generic #26~18.04.1-Ubuntu SMP Thu Aug 1 13:51:02 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
 - Output of `java -version`: OpenJDK 64-Bit GraalVM CE 19.1.1
 - Quarkus version or git rev: 0.21.1

**Additional context**
The list of extensions (features) used are: cdi, _hibernate-validator_, kubernetes-client, resteasy, resteasy-jackson, smallrye-health

I could provide additional logs if requested.

**Comment by geoand** [comment]
@gsmet is the SME here

**Comment by gsmet** [comment]
Interesting. I'll have a look.

**Comment by gsmet** [comment]
So I built the HV integration tests native executable again and the result is:
```
-rwxrwxr-x. 1 gsmet gsmet 31M Aug 20 17:13 target/quarkus-integration-test-hibernate-validator-999-SNAPSHOT-runner
```
which is in the ballpark of what I had in mind.

There's something wrong going on that is specific to your environment.

When building, you should have the following output:
```
Printing call tree to /data/home/gsmet/git/quarkus/integration-tests/hibernate-validator/target/reports/call_tree_quarkus-integration-test-hibernate-validator-999-SNAPSHOT-runner_20190820_171258.txt
Printing list of used classes to /data/home/gsmet/git/quarkus/integration-tests/hibernate-validator/target/reports/used_classes_quarkus-integration-test-hibernate-validator-999-SNAPSHOT-runner_20190820_171301.txt
Printing list of used packages to /data/home/gsmet/git/quarkus/integration-tests/hibernate-validator/target/reports/used_packages_quarkus-integration-test-hibernate-validator-999-SNAPSHOT-runner_20190820_171301.txt
```
Could you send me these files in both cases? It should just contain package names, class names and method names but if they contain sensitive information, feel free to redact them (consistently so I can still find the classes).

My email is gsmet (at) redhat (dot) com.

Thanks!

**Comment by jorsol** [comment]
Done! files send to you @gsmet 

If you need something else feel free to ask, I could try to help debugging if time allows me.

Thanks!

**Comment by gsmet** [comment]
So... the issue is using Hibernate Validator + Kubernetes Client.

Mostly all the fields in the Kubernetes API model are marked with `@Valid` which leads to the whole model being marked for reflection by HV and the whole model being included in the native image.

The model is huge thus leading to the native image being enormous.

FWIW, there is also another issue, there are a lot of classes defined such as:
```
+com.oracle.svm.reflect.WebhookClientConfig_url_c3d8a369a9370ceebed98ddbe3cfa2d492970e3e
+com.oracle.svm.reflect.Webhook_additionalProperties_82c1aeeefb0e1b8bed56255793d917a49cfc00e8
+com.oracle.svm.reflect.Webhook_admissionReviewVersions_469131ea2c8803ae56ccb9dd28596cb26dfc7150
+com.oracle.svm.reflect.Webhook_canEqual_55c378ebfe6114f5cf53af546011cc4d4031b373
+com.oracle.svm.reflect.Webhook_clientConfig_6dd9acbf343a9ef748e1660f4f23edfc99922820
+com.oracle.svm.reflect.Webhook_constructor_31b39507ebcca01ce7b2e384ff36977f9c4026d4
+com.oracle.svm.reflect.Webhook_constructor_f162977aa5ae19bc7a684b6b8996c7aa78b2f675
+com.oracle.svm.reflect.Webhook_equals_793ca0f77c24177b90ebac7434555664b13ab4d0
+com.oracle.svm.reflect.Webhook_failurePolicy_5d10d4c01c2bbf6c25e083d5c1517119345877f5
+com.oracle.svm.reflect.Webhook_getAdditionalProperties_6483074143e450486463b6ea8795710b40904166
+com.oracle.svm.reflect.Webhook_getAdmissionReviewVersions_5c22dc645d5918c2ac7a8707186e7f85ba0a00a7
+com.oracle.svm.reflect.Webhook_getClientConfig_ede7008a199e37d047740c6fbb5bf2dc2e039279
+com.oracle.svm.reflect.Webhook_getFailurePolicy_3d04c4772119fcb41a8d8160782bc2ff6f1336da
+com.oracle.svm.reflect.Webhook_getName_af50570f17d305bd13cd2b18718d1776c414c94c
+com.oracle.svm.reflect.Webhook_getNamespaceSelector_f15f75af9e3b4974ae20340ecd50610a09155a6c
+com.oracle.svm.reflect.Webhook_getRules_5acb0f43a86dd5cfc80a482d3fe9335404ca6825
+com.oracle.svm.reflect.Webhook_getSideEffects_6ffe5f17ca9295ea0822315d4655efd48b8a3535
+com.oracle.svm.reflect.Webhook_getTimeoutSeconds_62a669a8b64f5967e1e879d0637282affd8d37f7
+com.oracle.svm.reflect.Webhook_hashCode_c287b92d61453f8d67b23345e54b929799ce7a73
+com.oracle.svm.reflect.Webhook_name_a3b0ca36f128d263363813060baf5e7307e044b0
+com.oracle.svm.reflect.Webhook_namespaceSelector_8df07bb027d0dea8a2fdd3ec6f4d8c1ece1b5c73
+com.oracle.svm.reflect.Webhook_rules_3d5e6a460b831812d6a02216bc7b85a584025789
+com.oracle.svm.reflect.Webhook_setAdditionalProperty_9b80200d721f72d50720a0a0f9d1719628f987e9
+com.oracle.svm.reflect.Webhook_setAdmissionReviewVersions_e7b60cd4a4ac66c19ce4d1b7e2d735e80755c905
+com.oracle.svm.reflect.Webhook_setClientConfig_4d19237178ff4f6cca62310f8b6e255df1650165
+com.oracle.svm.reflect.Webhook_setFailurePolicy_75d518ed4d761bd18a01d69edd1a121095926b6b
+com.oracle.svm.reflect.Webhook_setName_6045827eb251ff0d0b7a17105b596078e5db3a1f
+com.oracle.svm.reflect.Webhook_setNamespaceSelector_5b1b83843aa99cfa294cbea3da5b90e9acc29247
+com.oracle.svm.reflect.Webhook_setRules_54a0f59f66b672f41e06348c0df7adb37a7663e6
+com.oracle.svm.reflect.Webhook_setSideEffects_dc2e74c4eabfb79603a3fceddd390c5dc4993616
+com.oracle.svm.reflect.Webhook_setTimeoutSeconds_d6a77302ebaf49389b57be07eea92e96342a8dd9
+com.oracle.svm.reflect.Webhook_sideEffects_bbe4985afd7bee9523a9cd6e665066d045b6fe3d
+com.oracle.svm.reflect.Webhook_timeoutSeconds_e313ded137a3e207c16ecc5fb50e9d588d18a2a7
+com.oracle.svm.reflect.Webhook_toString_e194793fdb6cc881f19c7333064a7ee216749538
```
basically one class per method and fields of the model classes (`Webhook` is a model class). That probably doesn't help either.

**Comment by geoand** [comment]
As discussed with @gsmet, one solution would be to use https://github.com/quarkusio/quarkus/blob/0.21.1/core/deployment/src/main/java/io/quarkus/deployment/builditem/substrate/ReflectiveClassBuildItem.java#L56.

Another solution might be for us to look into providing generated classes for libraries like Jackson and Hibernate Validator that are heavily used in third party libraries.
That way we could cut down on reflection

**Comment by gsmet** [comment]
We have an ongoing PR in HV but it predates Quarkus release so we were in stealth mode and implemented everything via ByteBuddy.

I will work on investigating porting it to Gizmo and see how it goes.

**Comment by geoand** [comment]
> We have an ongoing PR in HV but it predates Quarkus release so we were in stealth mode and implemented everything via ByteBuddy.
> 
> I will work on investigating porting it to Gizmo and see how it goes.

I'd be happe to help out with Gizmo (although my HV experience is limited)

**Comment by geoand** [comment]
@iocanel can you explain how the Kubernetes model used `@Valid` and if all it's instances are actually needed or could perhaps be trimmed down?
Asking because they are causing the native image produced by Quarkus to explode is size...

**Comment by iocanel** [comment]
The idea is that all Kubernetes model objects are validated by the builder itself (if a validation impl is present in the classpath).

AFAIR, this was added so that we can prevent users from creating instances containing illegal names, labels & annotations. This is why for such fields there are validation api annotations present in the model objects.

But I have absolutely no idea, why the @Valid annotation was added.
I can only guess that it was added to work around some issue ...

Let's try to remove it and see.

**Comment by geoand** [comment]
@iocanel do you want to do that or should I?

**Comment by gsmet** [comment]
Hmmm, the `@Valid` annotation is used to enable cascading so it might be useful in some cases. But right now, it looks as if it's added everywhere, ignoring if the bean might be constrained or not.

That being said, I'm not sure it's that easy to determine if a bean is contrained because you have to consider the entire tree of the bean (e.g. potentially the bean properties of the bean).

**Comment by geoand** [comment]
Actually, I'm having a tough time figuring out what is adding `@Valid`, @iocanel any pointers?

**Comment by iocanel** [comment]
`@Valid` is most probably added by the annotator module, that is passed to the jsonschema2pojo plugin.

**Comment by geoand** [comment]
@iocanel the only thing I see that might be relevant is: https://github.com/fabric8io/kubernetes-client/blob/1342a9f3b1db254b696ae6007d7eb04e32f78e54/kubernetes-model/kubernetes-model-annotator/src/main/java/io/fabric8/kubernetes/annotator/KubernetesTypeAnnotator.java#L77

**Comment by geoand** [comment]
@iocanel If I set `validationEnabled` to false, some tests no longer pass. So I think we need to look at this a little more

**Comment by iocanel** [comment]
yeah, that would completely disable validation, and I am not sure if it will remove the annotations or just skip calling the validator.

**Comment by geoand** [comment]
@iocanel ping me when you have some time so we can discuss please

**Comment by iocanel** [comment]
Ceated the PR on the client side, that will hopefully address the issue.

**Comment by geoand** [comment]
Awesome!
